File: backend/routers/relationships.py
from typing import Optional

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from db import get_db
from errors import ErrorCode, raise_coded
from dependencies import get_current_character
from models.character import Character
from models.relationship import RelationshipType
from schemas.block import BlockCreate, BlockListItem
from schemas.relationship import RelationshipCreate, RelationshipListItem
from services.block_service import block_character, list_blocks, unblock_character
from services.relationship_service import (
    block_relationship,
    build_relationship_item,
    create_relationship,
    list_relationships,
    unblock_relationship,
)

router = APIRouter()


# ---------------------------------------------------------------------------
# Blocks (ADR-0077) — their own record, addressed by character, never by edge.
# ---------------------------------------------------------------------------


@router.get("/blocks", response_model=list[BlockListItem])
async def list_my_blocks(
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> list[BlockListItem]:
    """List the characters the authenticated character has blocked.

    Outgoing only. A block is silent to the party it names (ADR-0077), so there
    is no route, field or error anywhere that reports an *incoming* one.
    """
    return await list_blocks(character_id=character.id, session=session)


@router.post("/blocks", response_model=BlockListItem, status_code=201)
async def block_character_route(
    data: BlockCreate,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> BlockListItem:
    """Block a character. No friend or foe declaration is needed or created.

    Idempotent — blocking someone already blocked answers the existing record.
    """
    return await block_character(
        blocker=character,
        character_id=data.character_id,
        session=session,
    )


@router.delete("/blocks/{character_id}", status_code=204)
async def unblock_character_route(
    character_id: int,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> None:
    """Lift a block. Only the character who created it can, and it restores
    nothing else — no friend or foe edge comes back with it (ADR-0077)."""
    await unblock_character(
        blocker=character,
        character_id=character_id,
        session=session,
    )


@router.get("", response_model=list[RelationshipListItem])
async def list_my_relationships(
    type: Optional[str] = None,
    status: Optional[str] = None,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> list[RelationshipListItem]:
    """List the authenticated character's outgoing friend and foe declarations,
    with the display status their pairing produces.

    Outgoing only, and blocks do not appear here at all — a block is its own
    record with its own route (ADR-0077).
    """
    return await list_relationships(
        character_id=character.id,
        session=session,
        type_filter=type,
        status_filter=status,
    )


@router.post("", response_model=RelationshipListItem, status_code=201)
async def create_relationship_route(
    data: RelationshipCreate,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> RelationshipListItem:
    """Declare a friend or foe relationship (instant, no pending state).

    Answers the enriched item, not the bare row (#1383): the display name,
    avatar, faction and derived ``display_status`` are what the profile draws,
    and it used to re-list every relationship the viewer holds to get them.
    """
    relationship = await create_relationship(
        from_character=character,
        to_character_id=data.to_character_id,
        rel_type=RelationshipType[data.type],
        session=session,
    )
    return await build_relationship_item(relationship, session)


@router.put("/{relationship_id}", response_model=RelationshipListItem)
async def block_relationship_route(
    relationship_id: int,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> RelationshipListItem:
    """Block the other party to this relationship, addressed by the edge's id.

    Superseded by `POST /relationships/blocks`, which takes a character id and
    needs no edge (ADR-0077). Kept for one release while the client moves;
    retires with #1907.
    """
    relationship = await block_relationship(
        relationship_id=relationship_id,
        character=character,
        session=session,
    )
    return await build_relationship_item(relationship, session)


@router.post("/{relationship_id}/unblock", response_model=RelationshipListItem)
async def unblock_relationship_route(
    relationship_id: int,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> RelationshipListItem:
    """Lift this caller's block on the other party to this relationship,
    addressed by the edge's id.

    Superseded by `DELETE /relationships/blocks/{character_id}` (ADR-0077).
    Kept for one release while the client moves; retires with #1907. Only the
    blocker's own record is deleted, and no edge changes.
    """
    relationship = await unblock_relationship(
        relationship_id=relationship_id,
        character=character,
        session=session,
    )
    return await build_relationship_item(relationship, session)


@router.delete("/{relationship_id}", status_code=204)
async def delete_relationship(
    relationship_id: int,
    character: Character = Depends(get_current_character),
    session: AsyncSession = Depends(get_db),
) -> None:
    """Remove a relationship. Only the declaring party can delete."""
    from models.relationship import Relationship

    relationship = await session.get(Relationship, relationship_id)
    if relationship is None:
        raise_coded(404, ErrorCode.relationship_not_found, "Relationship not found.")
    if relationship.from_character_id != character.id:
        raise_coded(
            403,
            ErrorCode.relationship_not_declarer,
            "Only the declaring party can delete a relationship.",
        )
    await session.delete(relationship)
    await session.flush()

File: frontend/src/api/relationships.ts
import { apiGet, apiPost, apiDelete } from './client'
import type { components } from './generated/schema'

/**
 * Matches backend RelationshipListItem (enriched list response).
 *
 * ponytail (#1400): `type` and `status` are narrower here than in the schema,
 * deliberately, so the four reads below narrow with a cast. The backend types
 * both as bare `str` (`schemas/relationship.py`) even though only these values
 * are ever written — `display_status` beside them shows what the annotated
 * version looks like. Widening these two to `string` would delete a guarantee
 * the UI branches on in order to match an annotation gap; the upgrade path is a
 * backend `StrEnum`, after which the casts go.
 */
export interface RelationshipListItem {
  id: number
  from_character_id: number
  to_character_id: number
  type: 'friend' | 'foe'
  /**
   * Dormant since ADR-0077: always `active`. A block is its own record now, so
   * nothing writes `blocked` here and nothing reads this field. The column
   * survives only because dropping a value from a PG enum costs a table rewrite
   * and buys a tidier enum.
   */
  status: 'active'
  created_at: string
  to_display_name: string
  to_avatar_url: string
  to_faction_slug: string
  // The reverse edge's raw type is not emitted: `display_status` is the
  // server's word on what the pair of edges means (#1387). No `Blocked` arm —
  // ADR-0077 stopped computing one.
  display_status: 'Mutual Friends' | 'Rivals' | 'Tsundere' | 'One-sided Friend' | 'One-sided Foe' | 'Secret Admirer' | 'Targeted' | 'Unknown'
}

interface RelationshipFilters {
  status?: string
  type?: string
}

export async function listRelationships(filters?: RelationshipFilters): Promise<RelationshipListItem[]> {
  const { data } = await apiGet('/relationships', { params: { query: filters } })
  return data as RelationshipListItem[]
}

// `createRelationship` answers the enriched item the list emits (#1383). It
// used to answer the bare row, so every caller re-ran `listRelationships()` and
// searched it for the edge it had just written.
//
// The item describes THE EDGE, not the viewer: `to_*` always names
// `to_character_id`.

export async function createRelationship(to_character_id: number, type: 'friend' | 'foe'): Promise<RelationshipListItem> {
  const { data } = await apiPost('/relationships', { body: { to_character_id, type } })
  return data as RelationshipListItem
}

export async function deleteRelationship(id: number): Promise<void> {
  await apiDelete('/relationships/{relationship_id}', {
    params: { path: { relationship_id: id } },
  })
}

// ---------------------------------------------------------------------------
// Blocks (ADR-0077) — their own record, addressed by CHARACTER, never by edge.
//
// This is why #1907 exists. While a block was a `status` on a friend/foe edge,
// blocking meant naming a row id, and the only ids the client ever held were
// its own outgoing edges — so a player foe'd by someone they had never declared
// anything about had nothing to act on, which is the one player the feature was
// written for. A character id is something the profile always has.
// ---------------------------------------------------------------------------

/** One character this viewer has blocked, enriched for the profile. */
export type BlockListItem = components['schemas']['BlockListItem']

/**
 * The characters this viewer has blocked. Outgoing only, always.
 *
 * There is no incoming counterpart to add here: a block is silent to the party
 * it names (ADR-0077 ruling 2), and the backend emits no route, field or error
 * that would report one. The blocked side's silence is therefore structural in
 * this client — there is nothing to render it from and no branch to get wrong.
 */
export async function listBlocks(): Promise<BlockListItem[]> {
  const { data } = await apiGet('/relationships/blocks')
  return data as BlockListItem[]
}

/**
 * Block a character. Needs no friend or foe declaration and creates none, and
 * blocking someone already blocked answers the existing record rather than a
 * conflict.
 */
export async function blockCharacter(character_id: number): Promise<BlockListItem> {
  const { data } = await apiPost('/relationships/blocks', { body: { character_id } })
  return data as BlockListItem
}

/**
 * Lift this viewer's own block. Only the blocker authored the record, so only
 * the blocker can delete one, and deleting it restores nothing else — no
 * friend/foe edge comes back with it (ADR-0077 ruling 5).
 */
export async function unblockCharacter(character_id: number): Promise<void> {
  await apiDelete('/relationships/blocks/{character_id}', {
    params: { path: { character_id } },
  })
}
